doc/adr/0014-synchronous_peer_discovery.md

```diff
@@ -0,0 +1,73 @@
+# 14. Synchronous Peer Discovery
+
+Date: 2023-04-19
+
+## Status
+
+Proposed
+
+## Context
+
+In order to increase security, it is desirable to observe only peers that are on the majority fork.
+This can be determined by querying discovered peers for their latest ordinal values prior to
+selecting one to observe. Afterwards, the majority ordinal can be determined and each peer is queried
+again for the `Signed[GlobalSnapshot]` corresponding to that ordinal. Then, the hashes of the
+snapshots are compared to determine the hash that is most prevelant among the list of Peers. A peer
+with that snapshot is then randomly chosen to be the one to be observed.
+
+Right now, the peer discovery process is a non-terminating process that runs alongside observation.
+It listens to a queue and processes peers that are put in that queue. Therefore, the list of known
+peers, at the time of observation, will only have 1 peer 99% of the time. In other words, determining
+the majority fork from among a list of peers that only has one entry is not useful. Therefore, it is
+imperative that there be multiple peers to choose from when choosing one to observe.
+
+To achieve this desired outcome, peer discovery can be run as a terminating process, after joining and
+before observation, in order to populate the list of known peers prior to the start of observation.
+However, peer discovery can be a long running process. Therefore, it is important that there continue
+to be non-terminsating peer discovery process that runs parrallel to other node processes.
+
+### The process
+
+The current bootstrap process will remain largely unchanged. A new node (_Node A_) will be given an
+existing node (_Node B_) to bootstrap off of. After starting, the two nodes will join each other.
+_Node B_ will be the first peer known by _Node A_. _Node A_ then queries _Node B_ for its list of
+known peers. These newly discovered peers are added to _Node A_'s list of known peers. The peer
+discovery process repeats for each peer added to the queue of discovered peers. This is what the
+current peer discovery process does.
+
+## Proposal
+
+A possible solution to our problem is to create a new separate, peer discovery process that will run
+after joining and before observation. It will not run as an asychronous process and will terminate
+upon hitting a terminating condition (e.g., a time limit, a certain number of peers discovered, etc.).
+It will also only run once, as part of the bootstrapping process.
+
+However, instead of creating a separate peer discovery process, we will simply delay the start of
+observation until a certain number of peers have been discovered.
```

Comment:
For the time being, the terminating conditions will be:

- at least 50 discovered peers
- 10 seconds has elapsed since the start of peer discovery.

Both conditions are subject to change as @ryle-goehausen and I continue to discuss this.